packages/vats/src/lib-chainStorage.js

```diff
@@ -0,0 +1,70 @@
+// @ts-check
+
+import { Far } from '@endo/far';
+
+const { details: X } = assert;
+
+// TODO: Formalize segment constraints.
+// Must be nonempty and disallow (unescaped) `.`, and for simplicity
+// (and future possibility of e.g. escaping) we currently limit to
+// ASCII alphanumeric plus underscore.
+const pathSegmentPattern = /^[a-zA-Z0-9_\-]{1,100}$/;
+
+/**
+ * Create a root storage node for a given backing function and root key.
+ *
+ * @param {(message: any) => any} toStorage a function for sending a storageMessage object to the storage implementation (cf. golang/cosmos/x/swingset/storage.go)
+ * @param {string} rootKey
+ */
+export function makeChainStorageRoot(toStorage, rootKey) {
+  assert.typeof(rootKey, 'string');
+
+  function makeChainStorageNode(key) {
+    const node = {
+      getKey() {
+        return key;
+      },
+      getChildNode(name) {
+        assert.typeof(name, 'string');
+        assert(
+          pathSegmentPattern.test(name),
+          X`Path segment must be a short ASCII identifier: ${name}`,
+        );
+        return makeChainStorageNode(`${key}.${name}`);
+      },
+      setValue(value) {
+        assert.typeof(value, 'string');
+        toStorage({ key, method: 'set', value });
+      },
+      async delete() {
+        assert(key !== rootKey);
+        // A 'set' with no value deletes a key if it has no children, but
```

Comment:
I don't think we _want_ to allow deleting a non-leaf node—IIUC, @warner and I agree that silent failure (by means of replacement with "set data to empty") seems like bad behavior that could leave orphans. But at any rate, this is a comment for #5385 rather than here.

Reply:
As I mentioned in https://github.com/Agoric/agoric-sdk/pull/5385#discussion_r876523826 the problem is solved and merged.  I think you and @warner may be misunderstanding the intended semantics:

- nodes with empty data are not stored in the tree, and unreferenced path linkage is cleared when emptying a node
- path linkage of nodes with nonempty data is always present, and when an empty node is filled any missing path linkage is built

The linkage information is just used to enumerate children.  That linkage information (pathToChildEntries) is kept in a separate place in the Merkle tree from the pathToData map.

Please remove this racy workaround.  It is unnecessary, and the main reason I'm requesting changes.